### PR TITLE
lib/osutil: IsDir returns true if path does not exist (fixes #3839)

### LINKED ISSUE #3839: Resync of offline machine stuck in "Out of Sync" state after dirs+files were added and then removed elsewhere
## Current Behavior
I am seeing syncthing get into a blocked "Out of Sync" state every time I switch branches in a code repo I have synchronized. This is a repo that I have been syncing for 6+ months without issue, but is now (with 0.14.16) failing. I cannot say for certain if 0.14.16 is the cause or if it was 0.14.14 or 15 as I haven't done a checkout of this particular branch for about a week prior to today. 

I currently see 27 items (a mix of files and directories) "unable to sync" due to a missing parent directory. Here is a sample screenshot from the UI with filenames scrambled.

screenshot from 2016-12-23 18-34-47

Looking at the paths the directories are all sub-directories of a (missing) parent directory, and the files are contained in those sub-directories, so that might be an indication of where the issue is. Also of note is that all of the indicated files and directories are not currently present on either machine. I had switched to a branch, then switched back while the failing machine was offline. The syncing machine state _should_ be a no-op actually because it's state matches that of the remote. I should note that I experienced the exact same thing on another machine and "fixed" it by switching branches directly on that machine.

## Expected Behavior
- Syncthing completes sync of all non-existent directories once offline machine is reconnected

I will try to get a reproduction script together, but it may take a little bit of time to pull off. Happy to run any suggested diagnostics on the failing machine. I seem to be running into all the syncthing bugs lately after 6+ months of worry-free use :-/

### Version Information

Syncthing Version: v0.14.16
OS Version: Arch Linux (x64)
Browser Version: N/A

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
The intent of the function is to verify that we are not traversing symlinks.
If the path along any component does not exist, we are not traversing anything, and it should be fine to let the updates through (as they are most likely only metadata updates)

/cc @Unrud

### CODE REVIEW

**Comment by calmh** [comment]
Note that this is a bugfix for the known cases when we are synchronizing a file in a directory that does not exist:

 - because the directory was deleted on one device, but the file updated on another
 - because the directory is ignored but the file inside it unignored

**Comment by calmh** [comment]
Should rename it if we do this check as the name is otherwise misleading.

**Comment by AudriusButkevicius** [comment]
Yeah I had like 3 minutes before I had to leave the house. Probably should be TraversesSymlink

File: lib/osutil/isdir.go [hunk 1]
```diff
@@ -14,13 +14,14 @@ import (
...
-// similar). Base and name must both be clean and name must be relative to
+// similar), but returns true if the path is inaccessible,
+// Base and name must both be clean and name must be relative to
 // base.
 func IsDir(base, name string) bool {
 	path := base
 	info, err := Lstat(path)
 	if err != nil {
-		return false
+		return os.IsNotExist(err)
```
Comment (calmh) [inline lib/osutil/isdir.go]: I don't think we should change this return. If the base directory doesn't exist we're in deep water, probably "folder path missing" etc.

File: lib/osutil/isdir.go [hunk 2]
```diff
@@ -14,13 +14,14 @@ import (
 
 // IsDir returns true if base and every path component of name up to and
 // including filepath.Join(base, name) is a directory (and not a symlink or
-// similar). Base and name must both be clean and name must be relative to
+// similar), but returns true if the path is inaccessible,
```
Comment (calmh) [inline lib/osutil/isdir.go]: This is broken. :) Maybe just `... is a directory or non-existent (and not a symlink`?

**Comment by calmh** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@calmh: Noted! Need another LGTM or explicit merge command.

**Comment by AudriusButkevicius** [comment]
Merge this or the other implementation, whatever you prefer.

**Comment by calmh** [comment]
@st-review merge it!

**Comment by st-review** [comment]
:ok_hand: Merged as 1a1e35d99872b86ec29d43aef61f9eac39de6b1a. Thanks, @AudriusButkevicius!

**Pull request closed without merging**